### ISSUE TITLE
Patter matching issue with GADTs

### ISSUE THREAD

**Issue by mhrde** [ocaml/ocaml#10463]
Hello,

I assume the following is a bug, but maybe its just a restriction of the compiler:
```ocaml
type t1 = C1 of int * int * string
type t2 = C2 of float

type 'a choose =
|   B1 : t1 choose
|   B2 : t2 choose

let func_ok : type t . (t choose * t) -> string = function
|   B1 , C1 (_, _, _)
|   B2 , C2 _             -> "dummy"

let func_bad : type t . (t * t choose) -> string = function
|   C1 (_, _, _) , B1
|   C2 _           , B2   -> "dummy"
```
In the code below, "func_ok" is compiled by the ocaml compiler, whereas "func_bad" is not, giving the following error
message:
```
File "tst.ml", line 12, characters 4-16:
12 | |   C1 (_, _, _) , B1
          ^^^^^^^^^^^
Error: This pattern matches values of type t1
       but a pattern was expected which matches values of type t
```
I came up with this issue while trying to understand the matching algorithm. I tried to formulate an example to prove, that
in "types.ml" of the ocaml compiler, the function "may_equal_constr" should also check the arity of "Constr_block" tags to
guarantee the equal column size invariant in the matching algorithm if different types are matched due to GADTs type selection.
If the bad function would pass the type checking unification (which it should IMHO), then this could also be an issue here,
because C1 and C2 have different arities, but they have the same constructor tag "Cstr_block 0", which is only checked by
"may_equal_constr".

Kind regards,

Michael Rose

**Comment by Octachron** [comment]
Patterns are type checked from left to right in presence of GADTs by design, see https://www.math.nagoya-u.ac.jp/~garrigue/papers/ml2011.pdf . 

Moreover, `may_equal_constr` has already been fixed to consider arity in #10295 .

**Comment by gasche** [comment]
I suspect that @Octachron hasn't had his morning coffee yet, or maybe too much of it; he may have forgotten to point out that it's a sensible question, and that your intuition on the importance of arity-checking was on point -- indeed we had a bug along these lines until not so long ago. To get arity-mismatch issues we have to build clauses that are incoherent (they contain constructors belonging to different type), and the fact that we decompose patterns in the same direction that we type-check GADT patterns (left to right) means that decompositions involving GADTs typically do not introduce incoherencies. But exceptions / extensible types do introduce inconsistencies, due to the "constructor ambiguity" problem (we may not know if two exception constructors are equal or not).

Thanks for the report, and let us know if you find any other interesting questions in this part of the codebase.
